sdk/core/core/test/cmocka/test_az_credential_client_secret.c

```diff
@@ -0,0 +1,227 @@
+    // returns 0.
+    assert_in_range(auth_attempt, 1, 2);
+
+    if (auth_attempt == 1)
+    {
+      response->_internal.http_response
+          = AZ_SPAN_FROM_STR("HTTP/1.1 500 Internal Server Error\r\n\r\n");
+    }
+    else
+    {
+      if (!redo_auth)
+      {
+        response->_internal.http_response
+            = AZ_SPAN_FROM_STR("HTTP/1.1 200 OK\r\n\r\n"
+                               "{ \"access_token\" : \"AccessToken\", \"expires_in\" : 3600 }");
+      }
+      else
+      {
+        response->_internal.http_response
+            = AZ_SPAN_FROM_STR("HTTP/1.1 200 OK\r\n\r\n"
+                               "{ \"access_token\" : \"NewAccessToken\", \"expires_in\" : 3600 }");
+      }
+    }
+  }
+  else // The actual HTTP request
+  {
+    bool has_auth_header = false;
+    int32_t const header_count = _az_http_request_headers_count(request);
+    for (int32_t i = 0; i < header_count; ++i)
+    {
+      az_pair header = { 0 };
+
+      az_result const ignore = az_http_request_get_header(request, i, &header);
+      (void)ignore;
+
+      if (az_span_is_content_equal(AZ_SPAN_FROM_STR("authorization"), header.key))
+      {
+        if (!redo_auth)
+        {
+          assert_true(
+              az_span_is_content_equal(AZ_SPAN_FROM_STR("Bearer AccessToken"), header.value));
+        }
+        else
+        {
+          assert_true(
+              az_span_is_content_equal(AZ_SPAN_FROM_STR("Bearer NewAccessToken"), header.value));
+        }
+
+        has_auth_header = true;
+      }
+    }
+
+    assert_true(has_auth_header);
+
+    static int attempt = 0;
+    ++attempt;
+
+    if (attempt == 1)
+    {
+      response->_internal.http_response = AZ_SPAN_FROM_STR("HTTP/1.1 200 OK\r\n\r\nResponse1");
+    }
+    else if (attempt == 2)
+    {
+      response->_internal.http_response = AZ_SPAN_FROM_STR("HTTP/1.1 200 OK\r\n\r\nResponse2");
+      redo_auth = true;
+    }
+    if (attempt == 3 && redo_auth)
+    {
+      response->_internal.http_response = AZ_SPAN_FROM_STR("HTTP/1.1 200 OK\r\n\r\nResponse3");
+    }
+    else if (attempt == 4 && redo_auth)
+    {
+      response->_internal.http_response = AZ_SPAN_FROM_STR("HTTP/1.1 200 OK\r\n\r\nResponse4");
+      redo_auth = false;
+    }
+  }
+
+  return AZ_OK;
+}
+
+#ifdef _az_MOCK_ENABLED
+az_result __wrap_az_http_client_send_request(_az_http_request* request, az_http_response* response);
+int64_t __wrap_az_platform_clock_msec();
+
+az_result __wrap_az_http_client_send_request(_az_http_request* request, az_http_response* response)
+{
+  return send_request(request, response);
+}
+
+int64_t __wrap_az_platform_clock_msec() { return (int64_t)mock(); }
+#endif // _az_MOCK_ENABLED
+
+int test_az_credential_client_secret()
+{
+  const struct CMUnitTest tests[] = {
+    cmocka_unit_test(test_credential_client_secret),
+  };
+  return cmocka_run_group_tests_name("az_core_credential_client_secret", tests, NULL, NULL);
```

Comment:
I am having difficulty understanding what this test is trying to validate. It seems more like an end-to-end test rather than a unit test. Is it possible to also add unit tests that test the behavior of individual components or methods as well?

Reply:
For example add unit test that only test the behavior of `_az_aad_build_url`, `_az_aad_build_body`, `_az_aad_request_token`, etc.

Reply:
I actually like this. This is much more thorough test than the one we had in the past, which did not caught a bug. This goes through all the flow realistically, making sure that we do retries, parse json, handle expiration, handle caching. And it does not invoke private functions like the previous one did - so you can change the internals, and as long as the implementation passes the test that is only using public/internal API, you know that you are good to go.

Reply:
I agree that e2e tests are valuable and we should certainly keep them. I am not suggesting replacing one with the other. However, if the test were to fail, the developer will have very little idea on what went wrong or where the bug is, requiring them to debug. That is why having granular unit tests help. Generally, a unit test should at least help guide/narrow down where the bug could potentially be.

https://www.guru99.com/unit-test-vs-integration-test.html
https://stackoverflow.com/questions/5357601/whats-the-difference-between-unit-tests-and-integration-tests